### PR TITLE
Adds full connection cycle capability to the Connection package.

### PR DESCRIPTION
**Pull request by zinigor**
#### Changes proposed in this Pull Request:
[x] Adds the register method to the Connection package.
[x] Hooks the register verification method for registration verification.
[x] Adds multisite registration capability.
[x] Makes sure the methods are moved with only the needed dependencies.
[x] Makes sure the example plugin is able to connect using only the dependent packages and minimal boilerplate code.

#### Is this a new feature or does it add/remove features to an existing part of Jetpack?
* This PR aims to make the Connection package usable without the Jetpack plugin itself.

#### Testing instructions:
* Use [the example client plugin](https://github.com/Automattic/client-example) on a clean WordPress site. It's bare bones, it has one administration page where it displays some debugging information, as well as lets you press a button to run the registration process. You should see that the site properly generates secrets, but can't proceed with user registration because handlers are not yet in place.

#### Proposed changelog entry for your changes:

* N/A

### CODE REVIEW

File: packages/connection/src/Manager.php
Comment (tyxla) [inline packages/connection/src/Manager.php]: Should we use the constants package for this one?

File: packages/connection/src/Manager.php
Comment (tyxla) [inline packages/connection/src/Manager.php]: What about this one? Are we planning to move it to a package?

File: jetpack.php
Comment (gravityrail) [inline jetpack.php]: This line disappearing is my favourite thing about this PR

File: packages/connection/src/Manager.php
Comment (zinigor) [reply packages/connection/src/Manager.php]: Yes, I'm hoping to merge Michael's #12717 into this PR once it gets merged.

File: packages/connection/src/Manager.php
Comment (zinigor) [reply packages/connection/src/Manager.php]: Yes, I'm hoping to merge Michael's #12717 into this PR once it gets merged.

**Comment by mdawaffe** [comment]
Example client plugin worked for me, and what I understand about the changes make sense :)

**Comment by mdawaffe** [comment]
491016a3f344e92d3f25bb28046d4dc52a51db70 moved the `$network_wpcom_blog_id = Jetpack_Options::get_option( 'id' );` to after the `switch_to_blog()`, so we're sending the subsite's wpcom_blog_id (which doesn't exist yet) in the subsite registration request. We should be sending the main site's wpcom_blog_id.

3ecdad75cf619f60dbb8c64294f7e133e7fb046f provides a sort of hacky fix (by unswitching/reswitching the blog).

File: packages/connection/src/Manager.php
Comment (mdawaffe) [inline packages/connection/src/Manager.php]: Typo.

```suggestion
		return apply_filters(
```

File: packages/connection/legacy/class.jetpack-xmlrpc-server.php
Comment (mdawaffe) [inline packages/connection/legacy/class.jetpack-xmlrpc-server.php]: This got refactored.

```suggestion
		Jetpack::connection()->require_jetpack_authentication();
```

File: packages/connection/src/Manager.php
Comment (mdawaffe) [inline packages/connection/src/Manager.php]: This code is namespaced now, so we need to be explicit here.
```suggestion
		if ( is_a( $user, '\WP_User' ) ) {
```

File: packages/connection/src/Manager.php
Comment (mdawaffe) [inline packages/connection/src/Manager.php]: …and here (namespace).
```suggestion
		return new \WP_User( $token_details['user_id'] );
```

File: packages/connection/src/Manager.php
Comment (zinigor) [reply packages/connection/src/Manager.php]: Thank you, I have found that one because of breaking tests too.

File: packages/connection/src/Manager.php
Comment (zinigor) [reply packages/connection/src/Manager.php]: Fixed that, thank you!

File: packages/connection/src/Manager.php
Comment (tyxla) [inline packages/connection/src/Manager.php]: Do we need a constructor if it doesn't do anything? Seems like we could remove it.

File: packages/connection/src/Manager.php
Comment (tyxla) [inline packages/connection/src/Manager.php]: Definitely not a requirement, but I like adding `@todo` blocks in the docblocks of the methods when there is something that needs refactoring. It would be nice to implement a better nonce verification here at some point.

File: packages/connection/src/Manager.php
Comment (tyxla) [inline packages/connection/src/Manager.php]: Could we use `return` at the end of the `if()` block and that way prevent from the need to have everything in an `else` block?

File: packages/connection/src/Manager.php
Comment (tyxla) [inline packages/connection/src/Manager.php]: 😅

File: packages/connection/src/Manager.php
Comment (tyxla) [inline packages/connection/src/Manager.php]: Do we have a plan to remove this one? Seems to be still coupling us to the plugin.

File: packages/connection/src/Manager.php
Comment (tyxla) [inline packages/connection/src/Manager.php]: Silly question perhaps, but do we need to do anything with the connector object references? Seems like we don't use them anywhere and are only interested in instantiating them.

File: packages/connection/src/Manager.php
Comment (tyxla) [inline packages/connection/src/Manager.php]: Given our current PHP min version support, I wonder if we could finally get rid of that hack.

File: packages/connection/src/Manager.php
Comment (tyxla) [inline packages/connection/src/Manager.php]: This looks a bit fragile. Could we have a whitelist of classes we allow here? I'm concerned anyone could hook a malicious class in the `wp_xmlrpc_server_class` filter that could do bad bad things.

File: packages/connection/src/Manager.php
Comment (tyxla) [inline packages/connection/src/Manager.php]: We need a double backslash here because of how they are parsed within single quotes.

```suggestion
		if ( is_a( $user, '\\WP_User' ) ) {
```

File: packages/connection/src/Manager.php
Comment (tyxla) [inline packages/connection/src/Manager.php]: Another case that we should correct to the actual current version we're working towards?

File: packages/connection/src/Manager.php
Comment (tyxla) [inline packages/connection/src/Manager.php]: Could also specify `@access private` for all those internal methods.

File: packages/connection/src/Manager.php
Comment (tyxla) [inline packages/connection/src/Manager.php]:
```suggestion
		add_action( 'pre_update_jetpack_option_register', array( '\\Jetpack_Options', 'delete_option' ) );
```

File: packages/connection/src/Manager.php
Comment (tyxla) [inline packages/connection/src/Manager.php]:
```suggestion
```

File: packages/connection/src/Manager.php
Comment (tyxla) [inline packages/connection/src/Manager.php]: Maybe we could specify where/when the error occurred? Like `An error occurred during attempt to validate remote registration`.

File: packages/connection/src/Manager.php
Comment (tyxla) [inline packages/connection/src/Manager.php]: Maybe we could specify more information about the error? Like `A Jetpack secret is required for validating registration` or something.

File: packages/connection/src/Manager.php
Comment (tyxla) [inline packages/connection/src/Manager.php]: Maybe we can be a bit more specific about the rules we want to ignore?

File: packages/connection/src/Manager.php
Comment (tyxla) [inline packages/connection/src/Manager.php]: Seems like we could ignore that i18n rule for the entire block of code, instead of doing it several times.

File: packages/connection/tests/php/test_Manager.php
Comment (tyxla) [inline packages/connection/tests/php/test_Manager.php]: Would be great to add some more tests (and remove tests that no longer make sense to be in the default Jetpack suite), but that's definitely not a requirement for this PR; we can do it in another PR.

File: tests/php/bootstrap.php
Comment (tyxla) [inline tests/php/bootstrap.php]: We should be able to use `__DIR__` here directly, since we no longer care about older PHP versions that didn't support it.

File: tests/php/general/test_class.jetpack-network.php
Comment (tyxla) [inline tests/php/general/test_class.jetpack-network.php]: Should we port this test somewhere, or is there a good reason we're removing it?

File: tests/php/general/test_class.jetpack-xmlrpc-server.php
Comment (tyxla) [inline tests/php/general/test_class.jetpack-xmlrpc-server.php]: 👍

File: tests/php/sync/test_class.jetpack-sync-callables.php
Comment (tyxla) [inline tests/php/sync/test_class.jetpack-sync-callables.php]: Should we use the `reset_saved_auth_state` from the connection manager here too?

**Review by tyxla — COMMENTED** [review]
Wow, this is amazing work, thank you for pulling it off! 

I'm yet to test it properly (will do it now), but I'm sending some feedback in the meantime. It's mostly minor stuff, really.

File: packages/connection/src/Manager.php
Comment (tyxla) [reply packages/connection/src/Manager.php]: FWIW I just hit a fatal here when trying the client example plugin:

```
[05-Aug-2019 10:50:04 UTC] PHP Warning:  Use of undefined constant JETPACK__PLUGIN_DIR - assumed 'JETPACK__PLUGIN_DIR' (this will throw an Error in a future version of PHP) in //wp-content/plugins/client-example/vendor/automattic/jetpack-connection/src/Manager.php on line 131
[05-Aug-2019 10:50:04 UTC] PHP Warning:  require_once(JETPACK__PLUGIN_DIRclass.jetpack-xmlrpc-server.php): failed to open stream: No such file or directory in //wp-content/plugins/client-example/vendor/automattic/jetpack-connection/src/Manager.php on line 131
[05-Aug-2019 10:50:04 UTC] PHP Fatal error:  require_once(): Failed opening required 'JETPACK__PLUGIN_DIRclass.jetpack-xmlrpc-server.php' (include_path='.:/opt/sp/php7.2/lib/php') in //wp-content/plugins/client-example/vendor/automattic/jetpack-connection/src/Manager.php on line 131
```

Maybe we should try to provide an `$xmlrpc_server` object before running this one? Also, we could totally use the `Constants` package for handling the constant, and I doubt we'll want to include a file from Jetpack, so we'll indeed need to make the xmlrpc server class independent.

Comment (tyxla) [reply packages/connection/src/Manager.php]: I was able to make this work by commenting this line, as we'll load the class as part of the connection package now:

```suggestion
			//require_once JETPACK__PLUGIN_DIR . 'class.jetpack-xmlrpc-server.php';
```

File: packages/connection/src/Manager.php
Comment (tyxla) [inline packages/connection/src/Manager.php]: I was able to make this work by telling Composer to look for it in the global namespace:
```suggestion
			$this->xmlrpc_server = new \Jetpack_XMLRPC_Server();
```

**Review by tyxla — COMMENTED** [review]
I made my way around this first error, see the comments:

File: packages/connection/src/Manager.php
Comment (tyxla) [inline packages/connection/src/Manager.php]: Got another warning here:

```
[05-Aug-2019 11:02:35 UTC] PHP Warning:  Use of undefined constant JETPACK__VERSION - assumed 'JETPACK__VERSION' (this will throw an Error in a future version of PHP) in /wp-content/plugins/client-example/vendor/automattic/jetpack-connection/src/Manager.php on line 1613
```

Should be resolved if we use the Constants package:

```suggestion
			'value'    => Constants::get_constant( 'JETPACK__VERSION' ),
```

File: class.jetpack-network.php
Comment (lezama) [reply class.jetpack-network.php]: it's probably already initialised at this point I guess?

**Comment by tyxla** [comment]
Jumping in to address feedback here and push it forward.

File: class.jetpack-network.php
Comment (tyxla) [reply class.jetpack-network.php]: I've verified and it's not possible; this is a serialized array, so it'll keep its integer type.

File: class.jetpack-network.php
Comment (tyxla) [reply class.jetpack-network.php]: Yes 👍

File: class.jetpack-network.php
Comment (tyxla) [reply class.jetpack-network.php]: Actually, the hook we hook this to is an action, so returning is not needed. I'll just change the `add_filter` to `add_action`.

File: class.jetpack-network.php
Comment (tyxla) [reply class.jetpack-network.php]: Going to keep it as-is for now, so we keep changes as minimal as possible.

File: class.jetpack.php
Comment (tyxla) [reply class.jetpack.php]: Let's leave that for another time - deprecating it will require updating tests, and I'd like to avoid making this PR larger. Adding a `@todo` for that.

File: class.jetpack.php
Comment (tyxla) [reply class.jetpack.php]: Will keep it as-is for now, as I'm not sure it should be changed. Seems like it's intentionally left here.

File: class.jetpack.php
Comment (lezama) [reply class.jetpack.php]: hmm, let's do it just in case, but probably no one is using them 🤷‍♂

File: packages/connection/legacy/class.jetpack-client.php
Comment (lezama) [reply packages/connection/legacy/class.jetpack-client.php]: let's delete this one :+1:

File: packages/connection/legacy/class.jetpack-xmlrpc-server.php
Comment (lezama) [reply packages/connection/legacy/class.jetpack-xmlrpc-server.php]: 👍

File: tests/php/general/test_class.jetpack-network.php
Comment (tyxla) [reply tests/php/general/test_class.jetpack-network.php]: No need to port it, we've removed the method.

File: class.jetpack.php
Comment (tyxla) [reply class.jetpack.php]: Done in 77069b4f2c317cd15dd48e960640b231b6625f35

File: packages/connection/src/Manager.php
Comment (tyxla) [reply packages/connection/src/Manager.php]: I think we're good to keep this as-is.

File: packages/connection/src/Manager.php
Comment (tyxla) [reply packages/connection/src/Manager.php]: Done in 0dd9046c77297db60bb895f86b190e148c5f9dcf

File: packages/connection/src/Manager.php
Comment (tyxla) [reply packages/connection/src/Manager.php]: Done in b5bc77afac0b66a68e75e2fbb22b2dba113c2fbe

File: packages/connection/src/Manager.php
Comment (tyxla) [reply packages/connection/src/Manager.php]: Added a `@todo` to address this in 5a4a72b6ac6e2610fb3f47041b762ae0e118ea60. It was like that before, so no rush to fix it now.

File: packages/connection/src/Manager.php
Comment (tyxla) [reply packages/connection/src/Manager.php]: Nope, this was indeed introduced in Jetpack 7.5.0.

File: packages/connection/src/Manager.php
Comment (tyxla) [reply packages/connection/src/Manager.php]: Not necessary for now.

File: packages/connection/src/Manager.php
Comment (tyxla) [reply packages/connection/src/Manager.php]: I think that's fine; I've also added a `@todo` block to audit whether those even make sense - it seems to me they should be using a textdomain, too. Auditing is not required now, and can be done separately.

File: packages/connection/src/Manager.php
Comment (tyxla) [reply packages/connection/src/Manager.php]: Committed that separately, marking as resolved.

File: packages/connection/src/Manager.php
Comment (tyxla) [reply packages/connection/src/Manager.php]: Removed 👍

File: packages/connection/src/Manager.php
Comment (tyxla) [reply packages/connection/src/Manager.php]: Added an error message here.

File: packages/connection/src/Manager.php
Comment (tyxla) [reply packages/connection/src/Manager.php]: Added an error message here.

File: packages/connection/src/Manager.php
Comment (tyxla) [reply packages/connection/src/Manager.php]: Done.

File: tests/php/bootstrap.php
Comment (tyxla) [reply tests/php/bootstrap.php]: Done in 34f7c850c5d03b48c82a4583f7e3daa6e837770c

File: tests/php/sync/test_class.jetpack-sync-callables.php
Comment (tyxla) [reply tests/php/sync/test_class.jetpack-sync-callables.php]: No, this will also cleanup the rest auth status, so let's leave it as-is for now.

**Comment by tyxla** [comment]
I think this is ready for another review!

More info on how to test it in p1HpG7-7lI-p2 and its comments.

**Review by lezama — DISMISSED** [review]
Things look to still connect and work, let's merge early and keep iterating and testing around this through the entire current cycle.

Great work @zinigor and @tyxla!

**Review by lezama — APPROVED** [review]
re approving after my last commit :)

**Pull request merged**
